### DISCUSSION TITLE
Session Data Mixing with Parallel User Access

### DISCUSSION THREAD

**Discussion by ibiscp** [fastapi/fastapi#10249] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import logging
from typing import List

from fastapi import APIRouter, HTTPException, WebSocket, WebSocketDisconnect
from fastapi.param_functions import Query

from jac.connectors.db import DB
from jac.connectors.models import Session
from jac.conversation import StreamingConversation
from jac.llm import OpenAILLM
from jac.stt import GoogleTranscriber
from jac.tts import GoogleSynthesizer

# Router
router = APIRouter()

# Logging
logging.basicConfig(level=logging.INFO, format="%(asctime)s [%(levelname)s] - %(message)s")
logger = logging.getLogger(__name__)


class ConnectionManager:
    def __init__(self) -> None:
        self.active_connections: List[WebSocket] = []
        self.active_conversations: dict[str, StreamingConversation] = {}

    async def connect(self, websocket: WebSocket, session_id: str, session: Session):
        await websocket.accept()
        self.active_connections.append(websocket)
        self.active_conversations[session_id] = StreamingConversation(
            sinthesizer=GoogleSynthesizer(session.bot.configuration),
            transcriber=GoogleTranscriber(),
            llm=OpenAILLM(session),
            websocket=websocket,
            session_id=session_id,
            session=session,
            session_time=session.session_time,
        )

    def get_conversation(self, session_id: str) -> StreamingConversation:
        if conversation := self.active_conversations.get(session_id):
            return conversation
        else:
            raise HTTPException(status_code=400, detail="Conversation not started!")

    def disconnect(self, websocket: WebSocket, session_id: str):
        self.active_connections.remove(websocket)
        self.active_conversations.pop(session_id, None)


manager = ConnectionManager()


@router.websocket("/conversation/{session_id}")
async def conversation(websocket: WebSocket, session_id: str, token: str = Query(...)):
    if not token:
        raise HTTPException(status_code=400, detail="Token is required!")

    # Connect to database
    db = DB(token=token)

    # Session and conversation
    session = db.get_session_by_id(session_id=session_id)
    await manager.connect(websocket, session.id, session)
    conversation_instance = manager.get_conversation(session.id)
    await conversation_instance.respond()

    try:
        while True:
            data = await websocket.receive_json()

            # Receive audio chunks from frontend
            # Calculate response
            # Reply answer to the frontend in chunks os sentence and audio

    except WebSocketDisconnect:
        logging.info("Conversation ended")

        # Save conversation in the database

        manager.disconnect(websocket, session_id)
```

### Description

I am currently working on a conversational application where the frontend is responsible for sending audio chunks to the backend. These chunks are then converted to text, processed to calculate an appropriate response, and sent back to the frontend.

While the application functions as expected during a single-user session, I am experiencing issues when multiple users access the system simultaneously. The issues manifest as mixed sessions where usernames are interchanged and session data is stored in incorrect database locations.

**Questions**

Has anyone encountered a similar issue while working on parallel user sessions?
Are there any known solutions or suggestions to prevent session mixing and ensure data integrity for multiple users accessing the system at the same time?

### Operating System

Linux

### FastAPI Version

0.95.1

### Pydantic Version

1.10.7

### Python Version

3.9

**Comment by n8sty** [comment]
I'm assuming that `jac` is some module you've developed yourself (maybe the name of your project) and not the `jac` that is available on PyPI. So, if that's true, without knowing more about what the `DB`  class is doing via an MRE you're not going to be able to get much help.

For reference, I've got a few FastAPI applications backed by a Postgres database that handle user sessions so this is almost definitely an implementation problem on your side.

**Reply by ibiscp** [reply]
Yes, for sure it is an implementation problem on my side, I just was not able to figure it out yet. Let me try to give more details, the `StreamingConversation` is a class that will handle the audio chunks, calculate an answer with the llm and synthesize this message back to the websocket.

It is not actually related to the database, the sessions are sometimes mixed together, for example the bot is talking to person A, and if a person B starts another chat, sometimes the bot will start calling person A as person B.

My question is if the solution presented above is able to separate the instances appropriately. Is there something I'm not seeing that is able to mix `StreamingConversation`?
